runtime-modules/bounty/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+// TODO: add max entries limit
+// TODO: benchmark all bounty creation parameters
+// TODO: add assertion for the created bounty object content
+
+/// pallet_bounty WeightInfo.
+/// Note: This was auto generated through the benchmark CLI using the `--weight-trait` flag
+pub trait WeightInfo {
+    fn create_bounty_by_council() -> Weight;
+    fn create_bounty_by_member() -> Weight;
+    fn cancel_bounty_by_member() -> Weight;
+    fn cancel_bounty_by_council() -> Weight;
+    fn veto_bounty() -> Weight;
+    fn fund_bounty() -> Weight;
+}
+
+type WeightInfoBounty<T> = <T as Trait>::WeightInfo;
+
+use frame_support::dispatch::{DispatchError, DispatchResult};
+use frame_support::traits::Currency;
+use frame_support::weights::Weight;
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, ensure, Parameter};
+use frame_system::ensure_root;
+use sp_arithmetic::traits::Saturating;
+use sp_arithmetic::traits::Zero;
+use sp_std::vec::Vec;
+
+use common::council::CouncilBudgetManager;
+use common::origin::MemberOriginValidator;
+use common::MemberId;
+
+use codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+/// Main pallet-bounty trait.
+pub trait Trait: frame_system::Trait + balances::Trait + common::Trait {
+    /// Events
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// Bounty Id type
+    type BountyId: From<u32> + Parameter + Default + Copy;
+
+    /// Validates member ID and origin combination.
+    type MemberOriginValidator: MemberOriginValidator<Self::Origin, MemberId<Self>, Self::AccountId>;
+
+    /// Weight information for extrinsics in this pallet.
+    type WeightInfo: WeightInfo;
+
+    /// Provides an access for the council budget.
+    type CouncilBudgetManager: CouncilBudgetManager<BalanceOf<Self>>;
+}
+
+/// Alias type for the BountyParameters.
+pub type BountyCreationParameters<T> = BountyParameters<
+    BalanceOf<T>,
+    <T as frame_system::Trait>::BlockNumber,
+    <T as common::Trait>::MemberId,
+>;
+
+/// Defines who will be the oracle of the work submissions.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Copy, Debug)]
+pub enum OracleType<MemberId> {
+    /// Specific member will be the oracle.
+    Member(MemberId),
+
+    /// Council will become an oracle.
+    Council,
+}
+
+impl<MemberId> Default for OracleType<MemberId> {
+    fn default() -> Self {
+        OracleType::Council
+    }
+}
+
+/// Defines who can submit the work.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum AssuranceContractType<MemberId> {
+    /// Anyone can submit the work.
+    Open,
+
+    /// Only specific members can submit the work.
+    Closed(Vec<MemberId>),
```

Comment:
Closed was originally intended to mean that only bounty creator can submit, and thus also get any cherry back if funding fails. In this version where any set of third parties can make entry, I assume creator still gets cherry back if funding fails?